Linting needs to check our sub go modules

**What steps did you take and what happened:**
1. Modify some code in a sub go module, e.g. test/infrastructure/docker, such that it needs to be `gofmt`-ed.
1. Run `make lint`
1. No errors reported

**What did you expect to happen:**
1. `gofmt` errors reported

**Anything else you would like to add:**
Apparently we need to invoke `golangci-lint` from the root of each go module:

1. project root (what we do now)
1. test/framework
1. test/infrastructure/docker

Unfortunately we can't just do
```
lint: $(GOLANGCI_LINT) ## Lint codebase
	$(GOLANGCI_LINT) run -v
	cd $(E2E_FRRAMEWORK_DIR); $(GOLANGCI_LINT) run -v
	cd $(CAPD_DIR); $(GOLANGCI_LINT) run -v
```
because `$(GOLANGCI_LINT)` evaluates to the relative path `hack/tools/bin/golangci-lint`, which breaks when you `cd`.

/kind bug
/priority important-soon
/milestone v0.3.0
/help

I'd suggest `Makefile`s in each directory that needs listing with their own lint targets. It's usually cleaner that way. That said, `abspath` does work:

```diff
$ git diff
diff --git a/Makefile b/Makefile
index dfc0aeea7..c34ee7345 100644
--- a/Makefile
+++ b/Makefile
@@ -46,7 +46,7 @@ RELEASE_NOTES := $(TOOLS_DIR)/$(RELEASE_NOTES_BIN)
 # Binaries.
 KUSTOMIZE := $(TOOLS_BIN_DIR)/kustomize
 CONTROLLER_GEN := $(TOOLS_BIN_DIR)/controller-gen
-GOLANGCI_LINT := $(TOOLS_BIN_DIR)/golangci-lint
+GOLANGCI_LINT := $(abspath $(TOOLS_BIN_DIR)/golangci-lint)
 CONVERSION_GEN := $(TOOLS_BIN_DIR)/conversion-gen
 
 # Bindata.
@@ -167,6 +167,8 @@ e2e-framework: ## Builds the CAPI e2e framework
 .PHONY: lint lint-full
 lint: $(GOLANGCI_LINT) ## Lint codebase
        $(GOLANGCI_LINT) run -v
+       cd $(E2E_FRAMEWORK_DIR); $(GOLANGCI_LINT) run -v
+       cd $(CAPD_DIR); $(GOLANGCI_LINT) run -v
 
 lint-full: $(GOLANGCI_LINT) ## Run slower linters to detect possible issues
        $(GOLANGCI_LINT) run -v --fast=false
```

```shell
$ make lint
/Users/akutz/Projects/go/src/sigs.k8s.io/cluster-api/hack/tools/bin/golangci-lint run -v
INFO [config_reader] Config search paths: [./ /Users/akutz/Projects/go/src/sigs.k8s.io/cluster-api /Users/akutz/Projects/go/src/sigs.k8s.io /Users/akutz/Projects/go/src /Users/akutz/Projects/go /Users/akutz/Projects /Users/akutz /Users /] 
INFO [config_reader] Used config file .golangci.yml 
INFO [lintersdb] Active 28 linters: [bodyclose deadcode depguard dogsled errcheck goconst gocritic gocyclo gofmt goimports golint gosec gosimple govet ineffassign interfacer maligned misspell nakedret prealloc scopelint staticcheck structcheck stylecheck typecheck unconvert unparam varcheck] 
INFO [loader] Go packages loading at mode 575 (compiled_files|exports_file|files|imports|name|deps|types_sizes) took 1.934607352s 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 35.931696ms 
INFO [runner/goanalysis_metalinter/goanalysis] analyzers took 2.373772639s with top 10 stages: buildssa: 605.423794ms, goimports: 238.333992ms, gofmt: 127.617244ms, misspell: 122.300591ms, unparam: 111.15789ms, golint: 92.369158ms, gosec: 81.593803ms, config: 74.812402ms, ineffassign: 68.662358ms, isgenerated: 64.243611ms 
WARN [runner] Can't run linter goanalysis_metalinter: SA1015: failed prerequisites: buildssa@sigs.k8s.io/cluster-api/controllers/metrics 
INFO [runner] processing took 4.9µs with stages: max_same_issues: 2.13µs, skip_dirs: 612ns, diff: 367ns, filename_unadjuster: 204ns, max_from_linter: 200ns, cgo: 195ns, skip_files: 190ns, nolint: 179ns, path_prettifier: 160ns, autogenerated_exclude: 145ns, identifier_marker: 137ns, exclude: 74ns, path_shortener: 64ns, max_per_file_from_linter: 63ns, uniq_by_line: 61ns, source_code: 61ns, exclude-rules: 58ns 
INFO [runner] linters took 1.995166802s with stages: goanalysis_metalinter: 1.995092769s 
INFO File cache stats: 47 entries of total size 337.0KiB 
INFO Memory: 41 samples, avg is 115.1MB, max is 204.8MB 
INFO Execution took 3.992317202s                  
cd test/framework; /Users/akutz/Projects/go/src/sigs.k8s.io/cluster-api/hack/tools/bin/golangci-lint run -v
INFO [config_reader] Config search paths: [./ /Users/akutz/Projects/go/src/sigs.k8s.io/cluster-api/test/framework /Users/akutz/Projects/go/src/sigs.k8s.io/cluster-api/test /Users/akutz/Projects/go/src/sigs.k8s.io/cluster-api /Users/akutz/Projects/go/src/sigs.k8s.io /Users/akutz/Projects/go/src /Users/akutz/Projects/go /Users/akutz/Projects /Users/akutz /Users /] 
INFO [config_reader] Used config file ../../.golangci.yml 
INFO [lintersdb] Active 28 linters: [bodyclose deadcode depguard dogsled errcheck goconst gocritic gocyclo gofmt goimports golint gosec gosimple govet ineffassign interfacer maligned misspell nakedret prealloc scopelint staticcheck structcheck stylecheck typecheck unconvert unparam varcheck] 
INFO [loader] Go packages loading at mode 575 (types_sizes|compiled_files|deps|name|exports_file|files|imports) took 1.99903612s 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 2.012377ms 
INFO [runner/goanalysis_metalinter/goanalysis] analyzers took 8.473336702s with top 10 stages: buildssa: 6.102266413s, fact_deprecated: 478.43509ms, ctrlflow: 399.116053ms, printf: 365.65601ms, fact_purity: 294.94662ms, unparam: 121.550343ms, goimports: 108.342259ms, interfacer: 97.236009ms, gofmt: 84.220621ms, inspect: 81.880609ms 
INFO [runner] Issues before processing: 42, after processing: 5 
INFO [runner] Processors filtering stat (out/in): exclude-rules: 5/5, uniq_by_line: 5/5, max_from_linter: 5/5, filename_unadjuster: 42/42, skip_files: 42/42, skip_dirs: 42/42, exclude: 5/42, cgo: 42/42, autogenerated_exclude: 42/42, identifier_marker: 42/42, max_per_file_from_linter: 5/5, path_prettifier: 42/42, diff: 5/5, path_shortener: 5/5, nolint: 5/5, max_same_issues: 5/5, source_code: 5/5 
INFO [runner] processing took 5.873217ms with stages: exclude: 1.611061ms, nolint: 1.532351ms, path_prettifier: 760.748µs, autogenerated_exclude: 735.156µs, identifier_marker: 655.393µs, source_code: 443.529µs, skip_dirs: 68.492µs, skip_files: 44.283µs, cgo: 11.259µs, uniq_by_line: 3.108µs, filename_unadjuster: 2.087µs, path_shortener: 2.007µs, max_per_file_from_linter: 1.698µs, max_same_issues: 1.142µs, diff: 430ns, exclude-rules: 264ns, max_from_linter: 209ns 
INFO [runner] linters took 2.612585777s with stages: goanalysis_metalinter: 2.606562359s 
config.go:263: File is not `gofmt`-ed with `-s` (gofmt)
				// Valid
			default:
				return errInvalidArg("Images[%d].LoadBehavior=%q", i, containerImage.LoadBehavior)
types.go:47: File is not `gofmt`-ed with `-s` (gofmt)

control_plane.go:99:34: Using the variable on range scope `obj` in function literal (scopelint)
			return mgmtClient.Create(ctx, obj)
			                              ^
management_cluster.go:95:26: Error return value of `imageLoader.LoadImage` is not checked (errcheck)
				imageLoader.LoadImage(ctx, image.Name)
				                     ^
control_plane.go:220:76: `opt` can be `sigs.k8s.io/controller-runtime/pkg/client.ListOption` (interfacer)
func ensureArtifactsDeleted(ctx context.Context, mgmtClient client.Client, opt *client.ListOptions) {
                                                                           ^
INFO File cache stats: 19 entries of total size 81.0KiB 
INFO Memory: 48 samples, avg is 195.4MB, max is 406.3MB 
INFO Execution took 4.645354632s                  
make: *** [Makefile:170: lint] Error 1
```

--- a/Makefile
+++ b/Makefile
@@ -44,13 +44,14 @@ RELEASE_NOTES_BIN := bin/release-notes
 RELEASE_NOTES := $(TOOLS_DIR)/$(RELEASE_NOTES_BIN)
 
 # Binaries.
-KUSTOMIZE := $(TOOLS_BIN_DIR)/kustomize
-CONTROLLER_GEN := $(TOOLS_BIN_DIR)/controller-gen
-GOLANGCI_LINT := $(TOOLS_BIN_DIR)/golangci-lint
-CONVERSION_GEN := $(TOOLS_BIN_DIR)/conversion-gen
+# Need to use abspath so we can invoke these from subdirectories
+KUSTOMIZE := $(abspath $(TOOLS_BIN_DIR)/kustomize)
+CONTROLLER_GEN := $(abspath $(TOOLS_BIN_DIR)/controller-gen)
+GOLANGCI_LINT := $(abspath $(TOOLS_BIN_DIR)/golangci-lint)
+CONVERSION_GEN := $(abspath $(TOOLS_BIN_DIR)/conversion-gen)
 
 # Bindata.
-GOBINDATA := $(TOOLS_BIN_DIR)/go-bindata
+GOBINDATA := $(abspath $(TOOLS_BIN_DIR)/go-bindata)
 GOBINDATA_CLUSTERCTL_DIR := cmd/clusterctl/config
 CERTMANAGER_COMPONENTS_GENERATED_FILE := cert-manager.yaml
 
@@ -176,9 +177,13 @@ e2e-framework: ## Builds the CAPI e2e framework
 .PHONY: lint lint-full
 lint: $(GOLANGCI_LINT) ## Lint codebase
 	$(GOLANGCI_LINT) run -v
+	cd $(E2E_FRAMEWORK_DIR); $(GOLANGCI_LINT) run -v
+	cd $(CAPD_DIR); $(GOLANGCI_LINT) run -v
 
 lint-full: $(GOLANGCI_LINT) ## Run slower linters to detect possible issues
 	$(GOLANGCI_LINT) run -v --fast=false
+	cd $(E2E_FRAMEWORK_DIR); $(GOLANGCI_LINT) run -v --fast=false
+	cd $(CAPD_DIR); $(GOLANGCI_LINT) run -v --fast=false
 
 ## --------------------------------------
 ## Generate / Manifests

--- a/test/framework/config.go
+++ b/test/framework/config.go
@@ -260,9 +260,9 @@ func (c *Config) Validate() error {
 		}
 		switch containerImage.LoadBehavior {
 		case MustLoadImage, TryLoadImage:
-				// Valid
-			default:
-				return errInvalidArg("Images[%d].LoadBehavior=%q", i, containerImage.LoadBehavior)
+			// Valid
+		default:
+			return errInvalidArg("Images[%d].LoadBehavior=%q", i, containerImage.LoadBehavior)
 		}
 	}
 	return nil

--- a/test/framework/control_plane.go
+++ b/test/framework/control_plane.go
@@ -93,7 +93,8 @@ func (input *ControlplaneClusterInput) ControlPlaneCluster() {
 	}, input.CreateTimeout, eventuallyInterval).Should(BeNil())
 
 	By("creating related resources")
-	for _, obj := range input.RelatedResources {
+	for i := range input.RelatedResources {
+		obj := input.RelatedResources[i]
 		By(fmt.Sprintf("creating a/an %s resource", obj.GetObjectKind().GroupVersionKind()))
 		Eventually(func() error {
 			return mgmtClient.Create(ctx, obj)
@@ -217,7 +218,7 @@ func (input *ControlplaneClusterInput) CleanUpCoreArtifacts() {
 	ensureArtifactsDeleted(ctx, mgmtClient, listOpts)
 }
 
-func ensureArtifactsDeleted(ctx context.Context, mgmtClient client.Client, opt *client.ListOptions) {
+func ensureArtifactsDeleted(ctx context.Context, mgmtClient client.Client, opt client.ListOption) {
 	// assertions
 	ml := &clusterv1.MachineList{}
 	Expect(mgmtClient.List(ctx, ml, opt)).To(Succeed())

--- a/test/framework/management_cluster.go
+++ b/test/framework/management_cluster.go
@@ -92,7 +92,7 @@ func InitManagementCluster(ctx context.Context, input *InitManagementClusterInpu
 				Expect(imageLoader.LoadImage(ctx, image.Name)).To(Succeed())
 			case TryLoadImage:
 				By(fmt.Sprintf("try to load image %s into the management cluster", image.Name))
-				imageLoader.LoadImage(ctx, image.Name)
+				imageLoader.LoadImage(ctx, image.Name) //nolint:errcheck
 			}
 		}
 	}

--- a/test/framework/types.go
+++ b/test/framework/types.go
@@ -44,4 +44,3 @@ type MachineDeployment struct {
 	BootstrapConfigTemplate runtime.Object
 	InfraMachineTemplate    runtime.Object
 }
-

